pkg/threescale/backend/manager.go

```diff
@@ -0,0 +1,387 @@
+func (b *Backend) Authorize(request threescale.Request) (threescale.AuthorizeResult, error) {
+	app, _, err := b.getApplicationFromRequest(request, request.Transactions[0])
+	if err != nil {
+		return nil, fmt.Errorf("unable to process request - %s", err.Error())
+	}
+
+	affectedMetrics := b.computeAffectedMetrics(app.LastResponse.GetHierarchy(), request.Transactions[0].Metrics)
+	_, isAuthorized := b.isAuthorized(app, affectedMetrics)
+	return AuthResponse{success: isAuthorized}, nil
+}
+
+// AuthRep authorizes a request based on the current cached values and reports
+// the new state to the cache in cases where the request has been authorized
+// If the request misses the cache, a remote call to 3scale is made
+// Request Transactions must not be nil and must not be empty
+// If multiple transactions are provided, all but the first is discarded
+func (b *Backend) AuthRep(request threescale.Request) (threescale.AuthorizeResult, error) {
+	app, cacheKey, err := b.getApplicationFromRequest(request, request.Transactions[0])
+	if err != nil {
+		return nil, fmt.Errorf("unable to process request - %s", err.Error())
+	}
+
+	affectedMetrics := b.computeAffectedMetrics(app.LastResponse.GetHierarchy(), request.Transactions[0].Metrics)
+	clonedApplication, isAuthorized := b.isAuthorized(app, affectedMetrics)
+	if isAuthorized {
+		b.cache.Set(cacheKey, clonedApplication)
+	}
+
+	return AuthResponse{success: isAuthorized}, nil
+}
+
+// Report the provided request transactions to the cache
+// Report always returns with 'Accepted' as bool 'true' and the error will only ever be non-nil
+// in cases where we have a cache miss and we also fail to get the hierarchy from 3scale from the resulting miss
+// Typically, we can handle cache miss and report to cache once we are aware of a hierarchy for the service
+// Supports multiple transactions
+func (b *Backend) Report(request threescale.Request) (threescale.ReportResult, error) {
+	var hierarchy api.Hierarchy
+	var err error
+
+	processingErr := make(chan error)
+
+	go func() {
+		for _, transaction := range request.Transactions {
+			// we support reporting in batches so for every transaction, grab the cache key and see if we
+			// have a match. if not, we can report locally regardless once we know the hierarchy
+			cacheKey := b.generateCacheKeyFromRequest(request, transaction)
+
+			app, ok := b.cache.Get(cacheKey)
+			if !ok {
+				// if we missed, configure a blank app
+				app = b.newApplication()
+			}
+
+			if hierarchy == nil {
+				if app.LastResponse == nil {
+					// while reporting locally, we don't care about cache misses, however we do need
+					// to deal with the exception of cases where we have no hierarchy
+					// which could lead to incorrect reports
+					app, err = b.handleCacheMiss(request, cacheKey)
+					processingErr <- err
+					if err != nil {
+						// we need to abort here and report back with the error
+						// we cant continue to report with bad state
+						return
```

Comment:
We have the data, so perhaps in the future if we can get a hierarchy we will be able to take these into account. (read: good to just drop it here for now, as dealing with this would add quite some complexity here we don't want now)